### ISSUE TITLE
Dreambooth example is not compatible with Windows

### ISSUE THREAD

**Issue by bbi-ai** [huggingface/diffusers#2041]
Same issue as described here: [https://github.com/huggingface/diffusers/issues/146](https://github.com/huggingface/diffusers/issues/1466) still occurring with main.
The fix from [https://github.com/huggingface/diffusers/pull/1547](https://github.com/huggingface/diffusers/pull/1547) does not completely fix it on Windows.

Windows 10
Python 3.10.9
diffusers 0.12.0.dev0 - pull from main 2023-01-19
multiprocess 0.70.14

When I manually edit train_dreambooth.py and remove parameter "num_workers=1," to torch.utils.data.DataLoader around line 654 everything works fine.

With the current code I get the following error:

```
Traceback (most recent call last):
  File "e:\Home\models\db\train_dreambooth.py", line 861, in <module>
    main(args)
  File "e:\Home\models\db\train_dreambooth.py", line 759, in main
    for step, batch in enumerate(train_dataloader):
  File "E:\Home\ds\venv\lib\site-packages\accelerate\data_loader.py", line 372, in __iter__
Traceback (most recent call last):
    dataloader_iter = super().__iter__()
  File "<string>", line 1, in <module>
  File "E:\Home\ds\venv\lib\site-packages\torch\utils\data\dataloader.py", line 435, in __iter__
  File "E:\anaconda3\envs\DS\lib\multiprocessing\spawn.py", line 116, in spawn_main
    return self._get_iterator()
    exitcode = _main(fd, parent_sentinel)
  File "E:\Home\ds\venv\lib\site-packages\torch\utils\data\dataloader.py", line 381, in _get_iterator
  File "E:\anaconda3\envs\DS\lib\multiprocessing\spawn.py", line 126, in _main
    return _MultiProcessingDataLoaderIter(self)
    self = reduction.pickle.load(from_parent)
  File "E:\Home\ds\venv\lib\site-packages\torch\utils\data\dataloader.py", line 1034, in __init__
EOFError: Ran out of input
    w.start()
  File "E:\anaconda3\envs\DS\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "E:\anaconda3\envs\DS\lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "E:\anaconda3\envs\DS\lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
  File "E:\anaconda3\envs\DS\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__
    reduction.dump(process_obj, to_child)
  File "E:\anaconda3\envs\DS\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'main.<locals>.<lambda>'
```

Command line:

```
accelerate launch train_dreambooth.py ^
  --pretrained_model_name_or_path=%MODEL_DIR%  ^
  --train_text_encoder ^
  --instance_data_dir=%INSTANCE_DIR% ^
  --class_data_dir=%CLASS_DIR% ^
  --output_dir=%OUTPUT_DIR% ^
  --with_prior_preservation --prior_loss_weight=1.0 ^
  --instance_prompt="......" ^
  --class_prompt="....." ^
  --resolution=512 ^
  --train_batch_size=1 ^
  --gradient_checkpointing ^
  --learning_rate=2e-6 ^
  --lr_scheduler="constant" ^
  --lr_warmup_steps=0 ^
  --checkpointing_steps=2000 ^
  --num_class_images=200 ^
  --max_train_steps=800
```

**Comment by patil-suraj** [comment]
Thanks for the issue. 

> When I manually edit train_dreambooth.py and remove parameter "num_workers=1," to torch.utils.data.DataLoader around line 654 everything works fine.

If `num_workers` is the issue, we could set it to 0 by default and let the users choose the value. Will send a fix soon.

**Comment by louisch** [comment]
Hi, apologies if you already know about this, but I'm fairly certain this behavior happens because on Windows, pytorch will pickle functions when using DataLoader with `num_workers` > 0, and with the built-in pickle module, closures cannot be pickled.

It's mentioned in the documentation here: https://pytorch.org/docs/stable/data.html#platform-specific-behaviors

It also mentioned in that segment to:
> Make sure that any custom collate_fn, worker_init_fn or dataset code is declared as top level definitions, outside of the __main__ check. This ensures that they are available in worker processes. (this is needed since functions are pickled as references only, not bytecode.)

So basically the example script just needs to be modified so that the offending function is defined at the top level somehow. That, or find some way to inject [dill](https://pypi.org/project/dill/) which is a third party pickle module that can pickle closures, into pytorch's code somehow...

I actually discovered this happened through running the LoRA example, actually (https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py), so it's a problem there too (when the parameter `dataloader_num_workers` for that script is set to greater than 0).
